lib/attrsets.nix

```diff
@@ -1726,6 +1728,75 @@ rec {
     if path == [] then "<root attribute path>"
     else concatMapStringsSep "." escapeNixIdentifier path;
 
+  /**
+    Turns a list of strings into a human-readable description of those
+    strings represented as an attribute path. The result of this function is
+    not intended to be machine-readable.
+    Create a new attribute set with `value` set at the nested attribute location specified in `attrPath`.
+
+    # Inputs
+
+    `context` (String)
+
+    : The error context in case a tag doesn't have a handler or the value is not valid.
+
+    `value` (AttrsOf Any)
+
+    : The value to match.
+
+    `handlers` (AttrsOf (Any -> Any))
+
+    : An attribute set containing a handler function for each case.
+
+    # Type
+
+    ```
+    matchAttrTag :: String -> AttrsOf Any -> AttrsOf (Any -> Any) -> Any
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.attrsets.matchAttrTag` usage example
+
+    ```nix
+    matchAttrTag "test" { n = 10; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => "It's the number 10"
+
+    matchAttrTag "test" { s = "Paul"; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => "It's the string Paul"
+
+    matchAttrTag "test" { unknown = null; } {
+      n = number: "It's the number ${toString number}";
+      s = str: "It's the string ${str}";
+    }
+    => error: test: No handler for tag "unknown"
```

Comment:
`"test"` doesn't seem like a great example. It'd help to have a more realistic use case. When it's a realistic example, it's easier to see what's expected from the context.

When this error occurs in the context of a module that defines an option of type `attrTag ...`, this case will only occur due to programming error, and I don't think the module author should have the foresight to clarify that in the context.

In other cases, you probably do want the context to be quite general.
Maybe we should have another function that takes an option instead of its value? Doesn't work for attrTags when used as part of type constructors like `attrsOf` though.

Reply:
What if the merged result of `attrTag` is the matching function itself :thinking: 

```nix
config.some.option {
  n = number: "It's the number ${toString number}";
  s = str: "It's the string ${str}";
}
```

This would be doable effectively in a backwards-compatible way using `__functor`! And this way we should be able to get a good error message for free. Not a big fan of `__functor` in general, but in this case I'm really wondering..

Reply:
For my use case, settings, I need to not have it:
```
       … while evaluating attribute '__functor'

error: cannot convert a function to JSON
```

Reply:
I guess it also wouldn't be very backwards compatible, because any code that asserts for the attribute set to only have one attribute would fail.